schemas/v3.1/meta/schema-object.json

```diff
@@ -0,0 +1,28 @@
+{
+    "$id": "https://spec.openapis.org/oas/3.1/meta/schema-object/2019-10",
+    "$schema": "https://json-schema.org/draft/2019-09/schema",
+    "$vocabulary": {
+        "https://json-schema.org/draft/2019-09/vocab/core": true,
+        "https://json-schema.org/draft/2019-09/vocab/applicator": true,
+        "https://json-schema.org/draft/2019-09/vocab/validation": true,
```

Comment:
@jdesrosiers the entire point of separating vocabularies and meta-schemas is to specify semantics with vocabularies and exact syntax with meta-schemas.

The semantics of the OAS 3.1 usage of these keywords are compatible with the standard semantics: if the syntax used matches the syntax for the standard semantics, then the standard semantics are correct (they are identical for both standard and OAS vocabularies).

If the syntax matches the old style, then the semantics come from the OAS vocabulary alone.

This is the correct approach, and is specifically one of the cases for which I designed the syntax/semantics separation and the notion of "compatible semantics".

Reply:
Honestly, this is not something I'd recommend doing, but we're shackled by Semantic Versioning into keeping this confusing compatibility.  Unless we decide that we could ignore semantic versioning on this point and drop the old boolean forms?  Pinging @darrelmiller in case semver conversations going on elsewhere move us in that direction.  This is really not an ideal thing to have but it's definitely better than not moving to a new draft.  If we could get away with dropping it I think that will be much less confusing as it will be harder to write a schema that doesn't work with standard validators.

Writing a plugin that handles this either-or thing isn't very clear- I might be pushing back on @jdesrosiers' suggestion here but he is 100% correct that this is confusing and weird, and that feedback is something the TSC should take into account.  I came up with this justification of either-or handling purely for OAI `exclusive*` semver-mandated compatibility purposes and would otherwise drop it.

Reply:
> the entire point of separating vocabularies and meta-schemas is to specify semantics with vocabularies and exact syntax with meta-schemas.

I understand that. What I'm arguing is that adding semantics _is_ changing the semantics from a vocabulary implementation point of view. I have to provide a different implementation for those 4 keywords than I do for draft 2019-09.

If it were the other way around with draft 2019-09 having extra functionality and OAS constraining it, there wouldn't be a problem. I could use the same implementation and just use the meta-schema to exclude certain usages.

> if the syntax used matches the syntax for the standard semantics, then the standard semantics are correct

That makes sense in words, but it's unhelpful from an implementation point of view. I still have to provide an alternate implementation even if some of the semantics overlap with the original.

Reply:
I did eventually come up with something I think could work to implement this as is, but it only serves to make things more complicated with no real benefit. I need to add a way for a vocabulary to explicitly define that one of it's keywords overrides that of another vocabulary. Then I need to rewrite the vocabulary loading functionality to make sure the right keyword implementations get loaded.

Even if there's a reasonable way to implement it this way, it seems harder to understand than it needs to be. By looking at the `$vocabulary` list, it's not clear that the "validation" vocabulary has been modified by the "extensions" vocabulary. If the OAS "validation" vocabulary has it's own identifier, it's clear that something is different.

Reply:
If not modifying the "validation" vocabulary is an option, that's highly preferred and makes this whole discussion moot.

Reply:
> By looking at the $vocabulary list, it's not clear that the "validation" vocabulary has been modified by the "extensions" vocabulary.

Because `$vocabulary` is intentionally underspecified so that we can get feedback on what is needed to complete the feature.

What is likely is that the `$vocabulary` URIs will identify some sort of machine-readable file that would say what keywords are part of the vocabulary, something about their classifications (does it act as an assertion?  does it provide an annotation, is it an in-place or child applicator? etc.), and something about what "semantic compatibility" means.

That is where an implementation could find out something like "if this keyword produces an error (not a validation failure, but an error) in the standard validation feature, it can alternatively be handled by this vocabulary."  Otherwise, the assumption is that a keyword has to work according to all vocabularies in use.

TBH, I'm hoping that by then OAS 4 is out and we can point to this sort of confusion and say "actually we're not going to support it."

There really isn't any substitute for putting something into production and letting people find out how it works or doesn't.  As you know better than most, 90% of this stuff is debated by the same 4 or 5 people in the JSON Schema repo, and at some point we have to admit that we're just guessing.

Anyway, thanks for raising all of this, it is exactly part of what "put it into production and see what happens" is supposed to raise.

Reply:
As things stand, full OAS 3.1 compatibility is far more important than eliminating an awkward corner of an underspecified feature that we have said numerous times we expect to tighten up before finalizing it.

Reply:
I'm not sure what you mean. Full OAS 3.1 compatibility is a given whether you take my suggestion or not. What am I missing?

Reply:
@jdesrosiers

> I'm not sure what you mean. Full OAS 3.1 compatibility is a given whether you take my suggestion or not. What am I missing?

I meant full JSON Schema compatibility in OAS 3.1- sorry, my mistake, and somehow I missed this earlier.  Anyway, see my other comment about hopefully being able to resolve this after all!